src/ray/rpc/node_manager/node_manager_client_pool.cc

```diff
@@ -0,0 +1,44 @@
+#include "ray/rpc/node_manager/node_manager_client_pool.h"
+
+namespace ray {
+namespace rpc {
+
+shared_ptr<ray::RayletClientInterface> NodeManagerClientPool::GetOrConnectByAddress(
+    const rpc::Address &address) {
+  RAY_CHECK(address.raylet_id() != "");
+  absl::MutexLock lock(&mu_);
+  auto raylet_id = NodeID::FromBinary(address.raylet_id());
+  auto it = client_map_.find(raylet_id);
+  if (it != client_map_.end()) {
+    return it->second;
+  }
+  auto connection = client_factory_(address);
+  // std::shared_ptr<NodeManagerWorkerClient>(NodeManagerWorkerClient::make(address.ip_address(),
+  // address.port(), client_call_manager_));
+  client_map_[raylet_id] = connection;
+
+  RAY_LOG(DEBUG) << "Connected to " << address.ip_address() << ":" << address.port();
+  return connection;
+}
+
+shared_ptr<ray::RayletClientInterface> NodeManagerClientPool::GetOrConnectByID(
+    ray::NodeID id) {
+  absl::MutexLock lock(&mu_);
+  auto it = client_map_.find(id);
+  if (it == client_map_.end()) {
+    return {};
```

Comment:
If we want to use this return, I think we should use the optional type. You can refer to `core_worker_client_pool` to see how to do that.